src/conversation_engine/engine.py

```diff
@@ -0,0 +1,409 @@
+
+        if self._handle_special_commands(text):
+            return
+
+        if self._state.state == ConversationState.IDLE:
+            # parse command and send_goal
+            self._handle_command(text)
+        elif self._state.state == ConversationState.WAIT_FOR_USER:
+            # Update semantics and send updated goal
+            self._handle_additional_info(text)
+        elif self._state.state == ConversationState.WAIT_FOR_ROBOT:
+            # User must wait for robot/action server to reply, cannot handle user input now
+            self._handle_user_while_waiting_for_robot(text)
+        elif self._state.state == ConversationState.ABORTING:
+            self._handle_user_while_aborting(text)
+
+    def _handle_special_commands(self, text):
+        """Check for special commands that should not be parsed further
+        @returns bool indicating if the text is a special command"""
+
+        stop_words = ["stop", "cancel", "quit", "reset"]
+        if any([word for word in stop_words if word in text]):
+            rospy.loginfo("_handle_special_commands('{}'):".format(text))
+            if self._state.state == ConversationState.IDLE:
+                self._robot_to_user_pub.publish(random.choice(["I'm not busy",
+                                                               "I'm not doing anything"]))
+                self._say_ready_for_command()
+            elif self._state.state == ConversationState.WAIT_FOR_ROBOT:
+                self._stop()
+            elif self._state.state == ConversationState.WAIT_FOR_USER:
+                self._robot_to_user_pub.publish(random.choice(["I'm waiting for you, there's nothing to stop",
+                                                               "I can't stop, you stop!"]))
+
+                self._start_new_conversation()
+            elif self._state.state == ConversationState.ABORTING:
+                self._robot_to_user_pub.publish(random.choice(["I'm already stopping, gimme some time!"]))
+            return True
+
+        kill_words = ["sudo kill"]
+        if any([word for word in kill_words if word in text]):
+            rospy.loginfo("_handle_special_commands('{}'): kill action server".format(text))
+
+            self._action_client.cancel_all_async()
+
+            self._robot_to_user_pub.publish(random.choice(["Woah, sorry dude!"]))
+            os.system("rosnode kill /state_machine")
+            self._robot_to_user_pub.publish(random.choice(["Killed the action_server, pray for resurrection"]))
+            self._start_new_conversation()  # This is assuming the state machine is back online when a command is received
+
+            return True
+
+    def _stop(self):
+        rospy.loginfo("_stop(): Cancelling goals, resetting state")
+
+        def hard_kill(event):
+            os.system("rosnode kill /state_machine")
+            self._robot_to_user_pub.publish(random.choice(["Headshot to the action_server, it had enough time to comply"]))
```

Comment:
It's not nice to show this to users (especially not when they're judging ;) )

Reply:
I know, but if you need this command, you're already in quite a pickle, better be funny about I guess. hard_kill is removed in a PR though.